### DISCUSSION TITLE
Access SqlConnections in Helper Class Serenity.NET 5

### DISCUSSION THREAD

**Discussion by rchawla-alphabyte** [serenity-is/Serenity#5790] [Help]
I'm migrating some code from previous version of Serenity to .NET CORE 5

I see the list of changes includes : "Replaced static SqlConnections with ISqlConnections abstraction, it is now theorically possible to use dynamic connection strings per request (multi tenancy++)"

How would I alter this method definition in a helper class to access SqlConnections? 
 Any Suggestions?

![image](https://user-images.githubusercontent.com/76563021/121271661-22d38500-c892-11eb-9ef9-7b18e8c5e258.png)

**Comment by minhhungit** [comment]
Hi @rchawla-alphabyte 
Check this post: https://www.davidezoccarato.cloud/resolving-instances-with-asp-net-core-di-in-static-classes/

**Reply by rchawla-alphabyte** [reply]
Hi @minhhungit I don't get what I would put in my DI Container method GetScope for ISqlConnections. ISqlConnections does not have GetRequiredService.

 public static IServiceScope GetScope(IServiceProvider serviceProvider = null)
        {
            var provider = serviceProvider ?? _serviceProvider;
            return provider?
                .GetRequiredService<IServiceScopeFactory>()
                .CreateScope();
        }

**Reply by minhhungit** [reply]
@rchawla-alphabyte 
Create a class `ServiceActivator` like this:
```csharp
/// <summary>
/// Add static service resolver to use when dependencies injection is not available
/// </summary>
public class ServiceActivator
{
  internal static IServiceProvider _serviceProvider = null;

  /// <summary>
  /// Configure ServiceActivator with full serviceProvider
  /// </summary>
  /// <param name="serviceProvider"></param>
  public static void Configure(IServiceProvider serviceProvider)
  {
      _serviceProvider = serviceProvider;
  }

  /// <summary>
  /// Create a scope where use this ServiceActivator
  /// </summary>
  /// <param name="serviceProvider"></param>
  /// <returns></returns>
  public static IServiceScope GetScope(IServiceProvider serviceProvider = null)
  {
      var provider = serviceProvider ?? _serviceProvider;
      return provider?
          .GetRequiredService<IServiceScopeFactory>()
          .CreateScope();
  }
}
```

and registry it in Startup.cs like this:

```csharp
public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILoggerFactory loggerFactory, IAntiforgery antiforgery)
{
    RowFieldsProvider.SetDefaultFrom(app.ApplicationServices);
    InitializeLocalTexts(app.ApplicationServices);
    ServiceActivator.Configure(app.ApplicationServices); // <============== add this line

    var reqLocOpt = new RequestLocalizationOptions();
```

Now you can use it to get ISqlConnections in your static class, this bellow is my example about getting list of users using static class

```csharp
public static class MyStaticClass
{
    public static List<Administration.Entities.UserRow> GetUsers()
    {
        using (var serviceScope = ServiceActivator.GetScope())
        {
            ISqlConnections sqlConnections = serviceScope.ServiceProvider.GetService<ISqlConnections>();
            using (var connection = sqlConnections.NewFor<Administration.Entities.UserRow>())
            {
                return connection.List<Administration.Entities.UserRow>();
            }
        }
    }
}
```
